### PR TITLE
switch to official jax optimiser library and fix loss function

### PR DESCRIPTION
**Pull request by drhutchinson99**
upated to work with newer versions of JAX and the
optax jax optimisation library

fixed a divide-by-zero in loss function resulting in NAN now matches pyTorch internal implementation from:
https://github.com/pytorch/pytorch/blob/c64e006fc399d528bb812ae589789d0365f3daf4/aten/src/ATen/native/Loss.cpp#L214

I am pretty new to this stuff so may I couldnt assess the error myself just managed to find and resolve :)

### CODE REVIEW

**Comment by ddehueck** [comment]
Thanks for playing around with project!
